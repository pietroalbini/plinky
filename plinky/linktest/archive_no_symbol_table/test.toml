archs = ["x86_64"]

[plinky.test]
cmd = ["${ar.foo}"]
kind = "link-fail"

[ar.foo]
output = "archive.a"
content = ["${asm.foo}"]
symbol-table = false

[asm.foo]
source = "foo.S"
