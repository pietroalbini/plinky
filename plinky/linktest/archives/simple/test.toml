archs = ["x86_64"]

[plinky.test]
cmd = ["${asm.entry}", "${ar.archived}"]
kind = "run-pass"
debug-print = ["loaded-object"]

[ar.archived]
output = "archived.a"
content = ["${asm.hello}", "${asm.goodbye}"]

[asm.entry]
source = "entry.S"

[asm.hello]
source = "hello.S"

[asm.goodbye]
source = "goodbye.S"
