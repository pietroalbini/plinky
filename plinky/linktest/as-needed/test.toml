archs = ["x86_64"]

[plinky.test]
cmd = ["${asm.entry}", "--as-needed", "${ld.used}", "${ld.unused}"]
kind = "link-pass"
debug-print = ["loaded-object=@inputs"]

[ld.used]
output = "libused.so"
content = ["${asm.used}"]
shared-library = true

[ld.unused]
output = "libunused.so"
content = ["${asm.unused}"]
shared-library = true

[asm.entry]
source = "entry.S"

[asm.used]
source = "used.S"

[asm.unused]
source = "unused.S"
