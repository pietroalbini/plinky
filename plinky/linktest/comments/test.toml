archs = ["x86_64"]

[plinky.test]
cmd = ["${asm.test}"]
kind = "link-pass"
debug-print = ["loaded-object=.comment", "layout", "final-elf=.comment"]

[asm.test]
source = "test.S"
