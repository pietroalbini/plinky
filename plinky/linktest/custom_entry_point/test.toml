archs = ["x86", "x86_64"]

[plinky.test]
cmd = ["${asm.hello}", "-e", "custom_entry"]
kind = "run-pass"
debug-print = ["final-elf=@meta"]

[asm.hello]
source = "hello.S"
