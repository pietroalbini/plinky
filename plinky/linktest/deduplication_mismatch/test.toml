archs = ["x86_64"]

[plinky.test]
cmd = ["${asm.foo}", "${asm.bar}"]
kind = "link-pass"

[asm.foo]
source = "foo.S"

[asm.bar]
source = "bar.S"
