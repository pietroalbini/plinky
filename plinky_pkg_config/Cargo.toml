[package]
name = "plinky_pkg_config"
version = "0.1.0"
edition = "2024"

[[test]]
name = "pkgtest"
path = "pkgtest/harness.rs"
harness = false

[dependencies]
plinky_macros.workspace = true

[dev-dependencies]
anyhow.workspace = true
insta.workspace = true
plinky_utils.workspace = true

[lints]
workspace = true
