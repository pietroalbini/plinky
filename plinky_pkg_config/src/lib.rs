#![feature(error_generic_member_access)]

mod discover;
mod lexer;
mod parser;
mod types;

pub use crate::discover::{discover, PkgConfigEnv};
pub use crate::lexer::LexError;
pub use crate::parser::ParseError;
pub use crate::types::*;
