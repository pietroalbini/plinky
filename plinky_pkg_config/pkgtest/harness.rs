#![feature(test)]

extern crate test;

use anyhow::Error;
use insta::{assert_snapshot, with_settings};
use plinky_pkg_config::PkgConfig;
use std::fs::read_to_string;
use std::path::{Path, PathBuf};
use std::process::exit;
use test::{test_main, ShouldPanic, TestDesc, TestDescAndFn, TestFn, TestName, TestType};

fn test(path: PathBuf) -> Result<(), Error> {
    let rendered = match PkgConfig::parse(&read_to_string(&path)?) {
        Ok(parsed) => {
            let PkgConfig {
                name,
                description,
                url,
                version,
                requires,
                requires_private,
                conflicts,
                cflags,
                libs,
                libs_private,
            } = parsed;

            let mut output = "Successfully parsed the file!\n\n".to_string();
            let mut push = |name: &str, slot: Option<String>| {
                if let Some(value) = slot {
                    output.push_str(&format!("{name}: {value}\n"));
                }
            };

            push("Name", name);
            push("Description", description);
            push("URL", url);
            push("Version", version);
            push("Requires", requires);
            push("Requires.private", requires_private);
            push("Conflicts", conflicts);
            push("CFlags", cflags);
            push("Libs", libs);
            push("Libs.private", libs_private);

            output
        }
        Err(err) => {
            format!("Failed to parse the file:\n\n{}", format_error(err.into()))
        }
    };

    let name = path.file_stem().unwrap().to_str().unwrap().to_string();
    with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => "",
    }, {
        assert_snapshot!(name, rendered);
    });

    Ok(())
}

fn gather(path: &Path) -> Result<Vec<TestDescAndFn>, Error> {
    let mut tests = Vec::new();
    for file in path.read_dir()? {
        let entry = file?.path();
        let name = entry.file_name().unwrap().to_str().unwrap().to_string();

        if entry.extension().and_then(|s| s.to_str()) == Some("pc") {
            tests.push(TestDescAndFn {
                desc: TestDesc {
                    name: TestName::DynTestName(name),
                    ignore: false,
                    ignore_message: None,
                    source_file: "",
                    start_line: 0,
                    start_col: 0,
                    end_line: 0,
                    end_col: 0,
                    should_panic: ShouldPanic::No,
                    compile_fail: false,
                    no_run: false,
                    test_type: TestType::IntegrationTest,
                },
                testfn: TestFn::DynTestFn(Box::new(move || match test(entry) {
                    Ok(()) => Ok(()),
                    Err(err) => panic!("{}", format_error(err)),
                })),
            });
        }
    }
    Ok(tests)
}

fn main() {
    let args = std::env::args().collect::<Vec<_>>();
    let path = Path::new(env!("CARGO_MANIFEST_DIR")).join("pkgtest");

    match gather(&path) {
        Ok(tests) => test_main(&args, tests, None),
        Err(err) => {
            eprintln!("{}", format_error(err));
            exit(1);
        }
    }
}

fn format_error(error: Error) -> String {
    let mut repr = format!("error: {error}\n");
    let mut source = error.source();
    while let Some(inner) = source {
        repr.push_str(&format!("  cause: {inner}\n"));
        source = inner.source();
    }
    repr
}
