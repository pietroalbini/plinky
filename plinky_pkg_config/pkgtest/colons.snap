---
source: plinky_pkg_config/pkgtest/harness.rs
expression: rendered
---
Successfully parsed the file!

Name: :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
